### PR TITLE
Adding max share and fee support

### PR DESCRIPTION
**Pull request by rrobison44**
Added new features:
- Added code that limits the maximum number of shares of each type that can be created.
- Added new account fee support.

[Jira CUS-1930](https://banno-jha.atlassian.net/browse/CUS-1930)

### CODE REVIEW

**Comment by tkainz** [comment]
Robbie…..

I’m seeing a lot of commits for this PR.  Should I wait until you quit updating it before doing the peer review or…?

**Comment by rrobison44** [comment]
Hi Tom.  You can review now.  I think I am mostly done with the updates.

Robby P. Robison | Staff Software Engineer | Digital Custom Development
[Logo, company name  Description automatically generated]



From: tkainz ***@***.***>
Date: Wednesday, November 16, 2022 at 3:09 PM
To: Banno/banno-powerons ***@***.***>
Cc: Robby Robison ***@***.***>, Mention ***@***.***>
Subject: Re: [Banno/banno-powerons] Adding max share and fee support (PR #149)

EXTERNAL EMAIL - Do not open attachments or click links from an unknown source. Forward suspicious emails to Anti-Spam.
Robbie…..

I’m seeing a lot of commits for this PR. Should I wait until you quit updating it before doing the peer review or…?

**Review by jucarson — CHANGES_REQUESTED** [review]
Robby, Three things besides the comments in the code... 

1. I believe you'll want to add the fee amount to the procedure CHECKSHAREIDS to remove a share type that has too high of a transfer amount AND fee amount so they can't pick it to begin with. 
2. Same with the max IDs... Should we just remove it from the list if they match that criteria? Seems crazy to do it both here and in the GETTERMSFUNDING state but could avoid errors.
3. Should all these changes be wrapped up inside the test account condition so it can be tested by certain users?

File: open-subshare/open-subshare-json-contract.md [hunk 1]
```diff
@@ -117,7 +117,22 @@ Returns the specific terms (if any) pertaining to the selected share type, the m
...
       }
-    ]
+    ],
+    "fee": {
+      "amount": "20.00",
+      "sourceAccounts": [
+        {
+          "name":"Savings",
+          "availableBalance": "1.00",
+          "memberAccountNumber": "1234567890S1234",
```
[the remark this answers is not in the document]
Comment (rrobison44) [reply open-subshare/open-subshare-json-contract.md:127]: Fixed

File: open-subshare/open-subshare-json-contract.md [hunk 2]
```diff
@@ -117,7 +117,22 @@ Returns the specific terms (if any) pertaining to the selected share type, the m
...
+        },
+        {
+          "name":"Savings",
+          "availableBalance": "1.00",
+          "memberAccountNumber": "1234567890S1234",
```
[the remark this answers is not in the document]
Comment (rrobison44) [reply open-subshare/open-subshare-json-contract.md:132]: Fixed

File: open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.POW [hunk 3]
```diff
@@ -515,9 +546,9 @@ DEFINE
 END
 
 SETUP
- PROGRAMVERSION="1.1.0"
- LASTMODDATE='10/20/22'
- LASTMODTIME="12:00 MT"
+ PROGRAMVERSION="1.2.0"
+ LASTMODDATE='11/16/22'
+ LASTMODTIME="17:00 MT"
  PROGRAMUPDATENOTE1="Copy name functionality added"
```
[the remark this answers is not in the document]
Comment (rrobison44) [reply open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.POW:552]: Done

File: open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.POW [hunk 4]
```diff
@@ -692,23 +725,32 @@ PRINT TITLE="BANNO.NEWSUBCREATE.V1.POW"
...
-         XFERSHAREFOUND=0
+         CALL GETSACFEEINFO   
+         IF SACFEESPECFILEERROR<>"" AND
+            SACFEESPECFILEERROR<>NOSUCHFILEERROR THEN
+          ERRORCODE=513
         END
 
-       IF VALUE(SACFUNDREQUIREFLAG)>0 AND
-          XFERSHAREFOUND=0 THEN
+       IF ERRORCODE<>0 THEN
```
[the remark this answers is not in the document]
Comment (rrobison44) [reply open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.POW:748]: I fixed this.  Added the error 506 back in.
```       IF ERRORCODE=0 THEN
        DO
         GETSHARETRANSFERRUN=TRANSFERRUNINITIAL
         CALL GETSHARETRANSFERS
         IF VALUE(SACFUNDREQUIREFLAG)>0 AND
            XFERSHAREFOUND=0 THEN
          ERRORCODE=506   
        END
```

File: open-subshare/open-subshare-json-contract.md [hunk 5]
```diff
@@ -117,7 +117,22 @@ Returns the specific terms (if any) pertaining to the selected share type, the m
...
+        }
+      ],
```
[the remark this answers is not in the document]
Comment (rrobison44) [reply open-subshare/open-subshare-json-contract.md:134]: Done

**Comment by rrobison44** [comment]
To address some of Justin's additional comments:

1. I believe you'll want to add the fee amount to the procedure CHECKSHAREIDS to remove a share type that has too high of a transfer amount AND fee amount so they can't pick it to begin with.  **[RR] I think it would get kind of messy if we added this to CHECKSHAREIDS.  We would basically have to run the fee PO against each share.  And since the electronic transfer share list is built in the initial state, I'm not sure how easy it would be to adjust the share list after a fee has been calculated in the GETTERMSFUNDING state.  Maybe we just need to add some additional error handling further down stream??**
2. Same with the max IDs... Should we just remove it from the list if they match that criteria? Seems crazy to do it both here and in the GETTERMSFUNDING state but could avoid errors.  **[RR] The design didn't specify that these share types that already equal or exceed the max should be excluded from the initial selection in the UX.  We are just supposed to display an error if they pick one that already at that max count.  I suppose we could run this by the BA to see if they want to change the design.**
3. Should all these changes be wrapped up inside the test account condition so it can be tested by certain users? **[RR]Not sure about this**

File: open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.POW [hunk 6]
```diff
@@ -515,9 +546,9 @@ DEFINE
...
 SETUP
- PROGRAMVERSION="1.1.0"
- LASTMODDATE='10/20/22'
- LASTMODTIME="12:00 MT"
+ PROGRAMVERSION="1.2.0"
+ LASTMODDATE='11/16/22'
+ LASTMODTIME="17:00 MT"
  PROGRAMUPDATENOTE1="Copy name functionality added"
  PROGRAMUPDATENOTE2="Error code output update, JSON code fix"
  CONFIGPROGRAMMINDATE='10/20/22'
```
[the remark this answers is not in the document]
Comment (jucarson) [reply open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.POW:554]: Sounds good

**Review by jucarson — COMMENTED** [review]
On the additional comments.. 

1. I see what you mean here, Maybe checking to be sure SACFEESHARECOUNT>0 at the end of the GETSACFEEINFO procedure would work? If not, throw an error 506 or create a new error. Looks like if there are no shares found, the JSON will just return an empty element and no errors are encountered but they won't be able to continue.

2. It's up to you if you want to check with the BA. Perhaps it makes more sense the way it is so the user knows why they can't open that type instead of it just being missing from the selection screen. 

3. Are there going to be changes to the UX so that a CU that's already live with an older version could test it using the Banno People permissions (i.e a different external application name)? If not, then all the changes should be conditional based on the TESTACCOUNT and TESTMODE flags and the old condition (for the last set of changes) should be removed. Looks like for this PowerOn's last changes, Tom just set the new variables to what they would be if the CU did not have the new parameter settings defined. Doing it that way makes it pretty easy.. See the end procedure SETTESTACCTSTATUS.

**Review by tkainz — CHANGES_REQUESTED** [review]
Robby.... just posting these notes I have so far. I'm still feeling - with what you're working on with Justin - that I'm aiming at a moving target

File: open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.FEES.V1.POW [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+PROCEDURE SACFEEINIT
+
+ SACFEECOUNT=0
+ SACFEEERROR=""
+END
+
+PROCEDURE SACFEEADD
+
+ IF SACFEECOUNT<SACFEEMAX THEN
```
[the remark this answers is not in the document]
Comment (rrobison44) [reply open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.FEES.V1.POW:144]: Done.  I went ahead and added but this is not an error we should ever see in reality.  That would be a lot of fees.  But it doesn't hurt to add it.

File: open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.CONFIG [hunk 8]
```diff
@@ -159,15 +168,17 @@ SETUP
  SACERRORCODEDEFAULTCHR(9)="Error getting new share rate"
  SACERRORCODEDEFAULTCHR(10)="Could not calculate new maturity date"
  SACERRORCODEDEFAULTCHR(11)="Error creating new share"
-
+ SACERRORCODEDEFAULTCHR(12)="No more shares of this type can be created"
```
[the remark this answers is not in the document]
Comment (rrobison44) [reply open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.CONFIG:171]: Added comment in the PowerOn section of the review.

File: open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.CONFIG [hunk 9]
```diff
@@ -1221,7 +1242,7 @@ PROCEDURE DISPLAYHTML
    HTMLVIEWLINE("<input id='groupName."+FORMAT("99",I)+"' type='text'"+
                 "size='30' maxlength='40' value='"+
                 SACSHARETYPEGROUPNAMES(I)+"'/>")
-   HTMLVIEWLINE("</td><td class='plsMin width50'>")
+   HTMLVIEWLINE("</td><td class='width100-'></td><td class='plsMin width50'>")
```
[the remark this answers is not in the document]
Comment (rrobison44) [reply open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.CONFIG:1245]: Fixed typo.

File: open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.CONFIG [hunk 10]
```diff
@@ -1234,11 +1255,30 @@ PROCEDURE DISPLAYHTML
...
+      END
+
      HTMLVIEWLINE("<td>")
-     HTMLVIEWLINE("ID Range:&nbsp;<input class='width600' id='shareIDs."+FORMAT("99",I)+"."+
+     HTMLVIEWLINE("ID Range:&nbsp;<input class='width500' id='shareIDs."+FORMAT("99",I)+"."+
                    FORMAT("999",J)+"' type='text' maxlength='80' ")
      HTMLVIEWLINE("onKeyPress='limitEntry(this,idRange)' value='")
-     HTMLVIEWLINE(SEGMENT(SACSHARETYPES(I,J),6,LENGTH(SACSHARETYPES(I,J)))+"' />")
+     HTMLVIEWLINE(IDRANGECHR+"' />")
+     HTMLVIEWLINE("<td>")
```
[the remark this answers is not in the document]
Comment (rrobison44) [reply open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.CONFIG:1277]: Looks like that was a bug in the original code.  Fixed.  Browser must handle it ok, though.

File: open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.CONFIG [hunk 11]
```diff
@@ -1271,27 +1311,27 @@ PROCEDURE INIT
 [* Build a list of valid (open) share types
 *]
  Q=CTRLCHR(34)
```
[the remark this answers is not in the document]
Comment (rrobison44) [reply open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.CONFIG:1313]: Done.

File: open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.CONFIG [hunk 12]
```diff
@@ -1416,8 +1460,23 @@ PROCEDURE PARSEHTML
         IF TEMPNUM>=0 AND TEMPNUM<=SACSHARETYPEGROUPSMAX AND
            TEMPNUM2>=0 AND TEMPNUM2<=SACSHARETYPESMAX AND
            TEMPLINE<>"" THEN
-          SACSHARETYPES(TEMPNUM,TEMPNUM2)=TEMPLINE+":"+
-                                          SACSHARETYPES(TEMPNUM,TEMPNUM2)
```
[the remark this answers is not in the document]
Comment (rrobison44) [reply open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.CONFIG:1420]: Yes, I believe this was a bug or code that was not required.  I may need to go back and figure out exactly what I was fixing.

File: open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.CONFIG [hunk 13]
```diff
@@ -159,15 +168,17 @@ SETUP
...
-    CALL PARSEHTML
-    CALL VALIDATEINPUT
-   END
+ VALIDINPUT=FALSE
+ WHILE VALIDINPUT=FALSE
+  DO
+   CALL DISPLAYHTML
+   CALL PARSEHTML
+   CALL VALIDATEINPUT
+  END
```
[the remark this answers is not in the document]
Comment (rrobison44) [reply open-subshare/REPWRITERSPECS/BANNO.NEWSUBCREATE.V1.CONFIG:180]: Done.

**Comment by rrobison44** [comment]
> On the additional comments..
> 
> 1. I see what you mean here, Maybe checking to be sure SACFEESHARECOUNT>0 at the end of the GETSACFEEINFO procedure would work? If not, throw an error 506 or create a new error. Looks like if there are no shares found, the JSON will just return an empty element and no errors are encountered but they won't be able to continue.
> 2. It's up to you if you want to check with the BA. Perhaps it makes more sense the way it is so the user knows why they can't open that type instead of it just being missing from the selection screen.
> 3. Are there going to be changes to the UX so that a CU that's already live with an older version could test it using the Banno People permissions (i.e a different external application name)? If not, then all the changes should be conditional based on the TESTACCOUNT and TESTMODE flags and the old condition (for the last set of changes) should be removed. Looks like for this PowerOn's last changes, Tom just set the new variables to what they would be if the CU did not have the new parameter settings defined. Doing it that way makes it pretty easy.. See the end procedure SETTESTACCTSTATUS.

[RR] (1) I implemented the SACFEESHARECOUNT>0 check as Justin suggested.  We will return error code 506 now in the condition where no fee shares are available to cover the amount of the fee. (2) No changes for now. (3) I still need to see how the test mode works and whether it is feasible in this situation.  But they can test the fees by just adding checks for account numbers in the fee specfile (i.e. only return a fee for the test accounts).  Also, for the share type create max, they can just set this max on a single group.

**Comment by rrobison44** [comment]
> Robby.... just posting these notes I have so far. I'm still feeling - with what you're working on with Justin - that I'm aiming at a moving target. Also... I couldn't seem to find any code which actually posts the calculated fees... where am I not looking? Ohhh. also, because these changes pertain to the "library" version, I think it'd be prudent to have as many looks as possible. Jordan is pretty good at catching a lot of small details.

[RR] Fixed issue with missing fee posting code.

**Review by jucarson — APPROVED** [review]
Looks like everything I found has been resolved and the TRANPERFORM for the fee is now in there. I approve, looks good. LMK if what you decide on the test accounts item.

**Pull request merged**